build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -842,6 +842,58 @@ message NodeProperties {
 
   // The UNIX file mode, e.g., 0755.
   google.protobuf.UInt32Value unix_mode = 3;
+
+  // Constrains server-side unification of identical filesystem nodes.
+  NodeUnificationConstraint.Value unification_constraint = 4;
+}
+
+// A `NodeUnificationConstraint` specifies whether servers are allowed to 
+// de-duplicate identical filesystem nodes when creating the input root.
+//
+// For performance reasons, it is useful for servers to perform de-duplication
+// on identical subsets of the input tree. For example, files with identical
+// content and properties might be hardlinked to reduce I/O, or separate
+// actions might have their inputs symlinked from a shared read-only blob cache.
+//
+// Such de-duplication can prevent correct execution of actions that inspect
+// file metadata, for example:
+//
+// * Tools that use `realpath` or `readlink()` to locate runtime data might
+//   fail if their executable is a symlink into a content-addressed cache.
+//
+// * Tools that identify files by inode rather than path might be confused
+//   if multiple input files have been mapped to a single inode.
+//
+// * Mutating writeable input files may cause unexpected effects on other
+//   parts of the input root.
+//
+// Node unification constraints allow the client to selectively restrict or
+// disable this optimization.
+message NodeUnificationConstraint {
+  enum Value {
+    // The server MAY use redirects, such as symlinks, to substitute for
+    // this node.
+    //
+    // For [FileNodes][build.bazel.remote.execution.v2.FileNode] and
+    // [DirectoryNodes][build.bazel.remote.execution.v2.DirectoryNode], the
+    // effective target MUST have the same type, content, and properties as
+    // the substituted node.
+    //
+    // For [SymlinkNodes][build.bazel.remote.execution.v2.SymlinkNode], the
+    // effective target MUST have the same type, content, and properties as
+    // the target of the substituted symlink.
+    ALLOW_REDIRECTS = 0;
+
+    // The server MAY use a filesystem alias, such as a hard link or bind
+    // mount, to substitute for this node.
```

Comment:
Hard links and bind mounts are absolutely not the same thing. As far as I know, bind mounts each have their own st_dev, meaning that files essentially don't alias. They each get their own identity.

Reply:
On Linux, files within bind mounts have the same `st_dev` and `st_ino` as the original.

```
$ uname -sr
Linux 5.14.0-1004-oem
$ mkdir srcdir dstdir
$ sudo mount -o bind srcdir dstdir
$ echo 'testing' > srcdir/test.txt
$ stat srcdir/test.txt 
  File: srcdir/test.txt
  Size: 8         	Blocks: 8          IO Block: 4096   regular file
Device: fd01h/64769d	Inode: 24259448    Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1000/    john)   Gid: ( 1000/    john)
[...]
$ stat dstdir/test.txt 
  File: dstdir/test.txt
  Size: 8         	Blocks: 8          IO Block: 4096   regular file
Device: fd01h/64769d	Inode: 24259448    Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1000/    john)   Gid: ( 1000/    john)
[...]
$ python3 -c 'import os; print(os.stat("srcdir/test.txt"))'
os.stat_result(st_mode=33204, st_ino=24259448, st_dev=64769, st_nlink=1, st_uid=1000, st_gid=1000, st_size=8, st_atime=1650338444, st_mtime=1650338444, st_ctime=1650338444)
$ python3 -c 'import os; print(os.stat("dstdir/test.txt"))'
os.stat_result(st_mode=33204, st_ino=24259448, st_dev=64769, st_nlink=1, st_uid=1000, st_gid=1000, st_size=8, st_atime=1650338444, st_mtime=1650338444, st_ctime=1650338444)
```